hubblestack/extmods/audit/grep.py

```diff
@@ -0,0 +1,160 @@
+# -*- encoding: utf-8 -*-
+"""
+Audit module for using grep to verify settings in files.
+
+Sample YAML data:
+
+.. code-block:: yaml
+
+    CIS-6.2.4:
+      grep.grep:
+        args:
+          - /etc/group
+        kwargs:
+          pattern: '^+:'
+          fail_on_match: True
+        description: Ensure no legacy "+" entries exist in /etc/group
+
+Required args/kwargs:
+
+    path
+        The absolute path of the file to match against
+    pattern
+        The pattern to use with the ``grep`` command.
+
+Optional kwargs:
+
+    grep_args
+        A list of args to pass to the ``grep`` command
+    fail_on_match
+        Defaults to False. If set to True, then if a match is found it will
+        count as a failure.
+    success_on_file_missing
+        Defaults to False. If set to True, then if a file is missing this check
+        will be marked as a success.
+    match_output
+        String to check for in the output of the grep command. If not provided,
+        any grep output will be considered a match.
+    match_output_regex
+        True/False. Whether to use regex when matching output. Defaults to
+        False.
+    match_output_multiline
+        True/False. Whether to use multiline flag for regex matching with
+        match_output_regex set to True. Defaults to True.
+"""
+from __future__ import absolute_import
+
+import logging
+import os
+import re
+
+from salt.exceptions import CommandExecutionError
+
+LOG = logging.getLogger(__name__)
+
+
+def grep(path,
+         pattern,
+         grep_args=None,
+         fail_on_match=False,
+         success_on_file_missing=False,
+         match_output=None,
+         match_output_regex=False,
+         match_output_multiline=True):
+    """
+    Use grep to match against the contents of a file.
+
+    :param path:
+        See module-level documentation
+    :param pattern:
+        See module-level documentation
+    :param grep_args:
+        See module-level documentation
+    :param fail_on_match:
+        See module-level documentation
+    :param success_on_file_missing:
+        See module-level documentation
+    :param match_output:
+        See module-level documentation
+    :param match_output_regex:
+        See module-level documentation
+    :param match_output_multiline:
+        See module-level documentation
+    :return:
+        Returns a tuple (success, {'grep_output': output}) where ``success``
+        is True or False based on the success of the check, and ``output`` is
+        the output of the ``grep`` command, for documentation purposes.
+    """
+    if not os.path.isfile(path):
+        if success_on_file_missing:
+            return True, {}
```

Comment:
Don't you need an

    else:
        return False, {}
I ran the salt-call cmd.run like it is in the code (I think), and got this:

    [root@hostn ~]# salt-call cmd.run 'grep hi /no/file' python_shell=False ignore_retcode=True
    local:
        grep: /no/file: No such file or directory
    [root@hostn ~]#

Reply:
Yes, I shall fix this.

Reply:
Done.